common/src/main/java/tachyon/security/PlainSaslServer.java

```diff
@@ -0,0 +1,151 @@
+ * 2.Write a factory class implements the SaslServerFactory
+ * 3.Write a JCA provider that registers the factory
+ */
+public class PlainSaslServer implements SaslServer {
+  private String mAuthcid;
+  private boolean mCompleted;
+  private CallbackHandler mHandler;
+
+  PlainSaslServer(CallbackHandler handler) throws SaslException {
+    mHandler = handler;
+  }
+
+  @Override
+  public String getMechanismName() {
+    return "PLAIN";
+  }
+
+  @Override
+  public byte[] evaluateResponse(byte[] response) throws SaslException {
+    if (mCompleted) {
+      throw new IllegalStateException("PLAIN authentication has completed");
+    }
+    if (response == null) {
+      throw new IllegalArgumentException("Received null response");
+    }
+    try {
+      // parse the response
+      // message   = [authzid] UTF8NUL authcid UTF8NUL passwd'
+      // authzid may be empty,then the authzid = authcid
+      String payload;
+      try {
+        payload = new String(response, "UTF-8");
+      } catch (Exception e) {
+        throw new IllegalArgumentException("Received corrupt response", e);
+      }
+      String[] parts = payload.split("\u0000", 3);
+      // validate response
+      if (parts.length != 3) {
+        throw new IllegalArgumentException("Received corrupt response");
+      }
+      String authzid = parts[0];
+      mAuthcid = parts[1];
+      String passwd = parts[2];
+      if (mAuthcid == null || mAuthcid.isEmpty()) {
+        throw new IllegalStateException("No authentication identity provided");
+      }
+      if (passwd == null || passwd.isEmpty()) {
+        throw new IllegalStateException("No password provided");
+      }
+      if (authzid == null || authzid.isEmpty()) { // authzid = authcid
+        authzid = mAuthcid;
+      }
+
+      NameCallback nameCallback = new NameCallback("User");
+      nameCallback.setName(mAuthcid);
+      PasswordCallback passwordCallback = new PasswordCallback("Password", false);
+      passwordCallback.setPassword(passwd.toCharArray());
+      AuthorizeCallback authCallback = new AuthorizeCallback(mAuthcid, authzid);
+
+      Callback[] cbList = {nameCallback, passwordCallback, authCallback};
+      mHandler.handle(cbList);
+      if (!authCallback.isAuthorized()) {
+        throw new SaslException("Authentication failed");
+      }
+    } catch (IOException ioe) {
+      throw new SaslException("Error validating the login", ioe);
+    } catch (UnsupportedCallbackException uce) {
+      throw new SaslException("Error validating the login", uce);
+    }
+    mCompleted = true;
+    return null;
+  }
+
+  @Override
+  public boolean isComplete() {
+    return mCompleted;
+  }
+
+  @Override
+  public String getAuthorizationID() {
+    throwIfNotComplete();
+    return mAuthcid;
+  }
+
+  @Override
+  public byte[] unwrap(byte[] incoming, int offset, int len) {
+    throwIfNotComplete();
+    throw new IllegalStateException("PLAIN supports neither integrity nor privacy");
```

Comment:
I am a bit confused here. This method (and also `wrap`) seems always throw `IllegalStateException` exceptions. Does that mean these two methods are not currently supported? If yes, is it better to throw `UnsupportedOperationException`? Or maybe you have plan to change them soon in later PRs, you can leave TODOs.